pkg/controller/transaction/controller.go

```diff
@@ -0,0 +1,228 @@
+						return false, err
+					}
+					return false, nil
+				}
+
+				return true, nil
+			}
+		}
+
+	}
+
+	//  If the transaction is in the Validating state, compute and validate the
+	//  Configuration for each target.
+	if transaction.Status.State == configapi.TransactionState_TRANSACTION_VALIDATING {
+		for _, change := range transaction.Changes {
+			modelName := utils.ToModelNameV2(change.TargetType, change.TargetVersion)
+			modelPlugin, ok := r.pluginRegistry.GetPlugin(modelName)
+			if !ok {
+				return false, errors.NewNotFound("model plugin not found")
+			}
+
+			pathValues := make([]*configapi.PathValue, len(change.Values))
+
+			for _, changeValue := range change.Values {
+				pathValue := &configapi.PathValue{
+					Path:  changeValue.Path,
+					Value: changeValue.Value,
+				}
+				pathValues = append(pathValues, pathValue)
+			}
+
+			jsonTree, err := tree.BuildTree(pathValues, true)
+			if err != nil {
+				return false, err
+			}
+			// If validation fails any target, mark the transaction Failed.
+			// If validation is successful, proceed to Applying.
+			err = modelPlugin.Validate(ctx, jsonTree)
+			if err != nil {
+				transaction.Status.State = configapi.TransactionState_TRANSACTION_FAILED
+				err = r.transactions.Update(ctx, transaction)
+				if err != nil {
+					if !errors.IsConflict(err) && !errors.IsNotFound(err) {
+						return false, err
+					}
+					return false, nil
+				}
+
+				return true, nil
+			}
+
+		}
+		transaction.Status.State = configapi.TransactionState_TRANSACTION_APPLYING
+		err := r.transactions.Update(ctx, transaction)
+		if err != nil {
+			if !errors.IsConflict(err) && !errors.IsNotFound(err) {
+				return false, err
+			}
+			return false, nil
+		}
+		return true, nil
+
+	}
+
+	// If the transaction is in the Applying state, update the Configuration for each
+	// target and Complete the transaction.
+	if transaction.Status.State == configapi.TransactionState_TRANSACTION_APPLYING {
+		if transaction.Atomic {
+			// TODO: Apply atomic transactions here
+		} else {
+			for _, transactionChange := range transaction.Changes {
+				config, err := r.configurations.Get(ctx, transactionChange.TargetID)
+				if err != nil {
+					if !errors.IsNotFound(err) {
+						return false, err
+					}
+					return false, nil
+				}
+				if config.Index < transaction.Index {
+					updatedConfigValues := make(map[string]*configapi.PathValue)
+					for targetID, pathValue := range config.Values {
+						pathValue.Index = transaction.Index
+						updatedConfigValues[targetID] = pathValue
+					}
+					config.Values = updatedConfigValues
+				}
+				config.Status.State = configapi.ConfigurationState_CONFIGURATION_PENDING
+				config.Status.TransactionIndex = transaction.Index
+				err = r.configurations.Update(ctx, config)
+				if err != nil {
+					if !errors.IsConflict(err) && !errors.IsNotFound(err) {
+						return false, err
+					}
+					return false, nil
+				}
```

Comment:
Return `true` here too. This will become unnecessary if we move each state into a separate method though